mmap-alloc/src/lib.rs

```diff
@@ -411,90 +283,199 @@ impl MapAlloc {
+        let new_size = next_multiple(new_layout.size(), self.pagesize);
+        if old_size == new_size {
+            return Ok(ptr);
+        }
+        remap(ptr, old_size, new_layout.size(), false).ok_or(AllocErr::Exhausted {
+            request: new_layout,
+        })
+    }
+
+    #[cfg(any(target_os = "macos", windows))]
+    unsafe fn realloc(
+        &mut self,
+        ptr: *mut u8,
+        layout: Layout,
+        new_layout: Layout,
+    ) -> Result<*mut u8, AllocErr> {
+        // Adapted from the Rust standard library.
+
+        debug_assert!(
+            layout.size() > 0 && new_layout.size() > 0,
+            "realloc: size of layout and new_layout must be non-zero"
+        );
+
         // alignment less than a page is fine because page-aligned objects are also aligned to
         // any alignment less than a page
         if new_layout.align() > self.pagesize {
-            return Err(AllocErr::invalid_input("cannot support alignment greater than a page"));
+            return Err(AllocErr::invalid_input(
+                "cannot support alignment greater than a page",
+            ));
         }
-        debug_assert!(layout.size() > 0 && new_layout.size() > 0,
-                      "usable_size: size of layout and new_layout must be non-zero");
 
         let old_size = next_multiple(layout.size(), self.pagesize);
         let new_size = next_multiple(new_layout.size(), self.pagesize);
+
         if old_size == new_size {
             return Ok(ptr);
         }
-        self.realloc_helper(ptr, old_size, new_layout.size()).ok_or(AllocErr::Exhausted { request: new_layout })
+
+        if !cfg!(windows) {
+            // Windows cannot grow or shrink existing mappings.
+            if new_size >= old_size {
+                if let Ok(()) = self.grow_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            } else if new_size < old_size {
+                if let Ok(()) = self.shrink_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            }
+        }
+
+        // otherwise, fall back on alloc + copy + dealloc.
+        let result = Alloc::alloc(self, new_layout);
+        if let Ok(new_ptr) = result {
+            use core::cmp;
+
+            // If we're configured to allocate un-readable or un-writable memory, we need to
+            // temporarily mark the old allocation readable and the new allocation writable in
+            // order to be able to copy the data.
+            let fix_permissions = !self.read || !self.write;
+            if fix_permissions {
+                protect(ptr, old_size, perms::get_perm(true, self.write, self.exec));
+                protect(
+                    new_ptr,
+                    new_size,
+                    perms::get_perm(self.read, true, self.exec),
+                );
+            }
```

Comment:
I'd prefer to split this, so if reads are allowed, but not writes, there's no reason to fix permissions on the old allocations, only the new one, to allow writing, then reset them after. Likewise, for the opposite, if writes are allowed, but not reads, we only have to fix permissions on the old allocation, and never have to reset them back.

Reply:
Again, missing important context in the short preview. Starts [here](https://github.com/ezrosent/allocators-rs/pull/87/files/3e5636db556f92e292af5ece2a13e21ab4809f89#diff-7be4d90297df3fdf7c0f62394837b6d5R444)

Reply:
Fixed.